rust/src/tx_builder.rs

```diff
@@ -0,0 +1,164 @@
+use super::*;
+use super::fees;
+use super::utils;
+use std::collections::BTreeSet;
+
+// We need to know how many of each type of witness will be in the transaction so we can calculate the tx fee
+struct MockWitnessSet {
+  vkeys: BTreeSet<AddrKeyHash>,
+  scripts: BTreeSet<ScriptHash>,
+  bootstraps: BTreeSet<String>,
+}
+
+#[wasm_bindgen]
+pub struct TransactionBuilder {
+    inputs: TransactionInputs,
+    outputs: TransactionOutputs,
+    fee: Option<Coin>,
+    ttl: Option<u32>, // absolute slot number
+    certs: Option<Certificates>,
+    withdrawals: Option<Withdrawals>,
+    metadata: Option<TransactionMetadata>,
+    input_types: MockWitnessSet,
+}
+
+// comes from witsVKeyNeeded in the Ledger spec
+fn witness_keys_for_cert(cert_enum: &Certificate, keys: &mut BTreeSet<AddrKeyHash>) {
+    match &cert_enum.0 {
+        CertificateEnum::StakeRegistration(cert) => {
+            keys.insert(cert.stake_credential().to_keyhash().unwrap());
+        },
+        CertificateEnum::StakeDeregistration(cert) => {
+            keys.insert(cert.stake_credential().to_keyhash().unwrap());
+        },
+        CertificateEnum::StakeDelegation(cert) => {
+            keys.insert(cert.stake_credential().to_keyhash().unwrap());
+        },
+        CertificateEnum::PoolRegistration(cert) => {
+            for owner in &cert.pool_params().pool_owners().0 {
+                keys.insert(owner.clone());
+            }
+            keys.insert(
+                AddrKeyHash::from_bytes(cert.pool_params().operator().to_bytes()).unwrap()
+            );
+        },
+        CertificateEnum::PoolRetirement(cert) => {
+            keys.insert(
+                AddrKeyHash::from_bytes(cert.pool_keyhash().to_bytes()).unwrap()
+            );
+        },
+        CertificateEnum::GenesisKeyDelegation(cert) => {
+            keys.insert(
+                AddrKeyHash::from_bytes(cert.genesis_delegate_hash().to_bytes()).unwrap()
+            );
+        },
+        CertificateEnum::MoveInstantaneousRewardsCert(_cert) => {},
+    }
+}
+
+#[wasm_bindgen]
+impl TransactionBuilder {
+    // We have to know what kind of inputs these are to know what kind of mock witnesses to create since
+    // 1) mock witnesses have different lengths depending on the type which changes the expecting fee
+    // 2) Witnesses are a set so we need to get rid of duplicates to avoid over-estimating the fee
+    pub fn add_key_input(&mut self, hash: &AddrKeyHash, input: &TransactionInput) {
+        self.inputs.add(input);
+        self.input_types.vkeys.insert(hash.clone());
+    }
+    pub fn add_script_input(&mut self, hash: &ScriptHash, input: &TransactionInput) {
+        self.inputs.add(input);
```

Comment:
I'm not really sure how script witnesses work and if this approach even works for script witnesses to be honest. We won't need them for the initial release though so maybe it's okay.